lib/ui/dart_ui.cc

```diff
@@ -34,70 +34,278 @@
+  V(SceneBuilder, pushOffset, 5)                       \
+  V(SceneBuilder, pushOpacity, 6)                      \
+  V(SceneBuilder, pushPhysicalShape, 8)                \
+  V(SceneBuilder, pushShaderMask, 10)                  \
+  V(SceneBuilder, pushTransformHandle, 4)              \
+  V(SceneBuilder, setCheckerboardOffscreenLayers, 2)   \
+  V(SceneBuilder, setCheckerboardRasterCacheImages, 2) \
+  V(SceneBuilder, setRasterizerTracingThreshold, 2)    \
+  V(Scene, dispose, 1)                                 \
+  V(Scene, toImage, 4)                                 \
+  V(Scene, toImageSync, 4)                             \
+  V(SemanticsUpdateBuilder, build, 2)                  \
+  V(SemanticsUpdateBuilder, updateCustomAction, 5)     \
+  V(SemanticsUpdateBuilder, updateNode, 36)            \
+  V(SemanticsUpdate, dispose, 1)
 
-}  // namespace
+#define FFI_FUNCTION_MATCH(FUNCTION, ARGS)                                 \
+  if (strcmp(name, #FUNCTION) == 0 && args == ARGS) {                      \
+    return reinterpret_cast<void*>(                                        \
+        tonic::FfiDispatcher<void, decltype(&FUNCTION), &FUNCTION>::Call); \
+  }
 
-void DartUI::InitForGlobal() {
-  if (!g_natives) {
-    g_natives = new tonic::DartLibraryNatives();
-    Canvas::RegisterNatives(g_natives);
-    CanvasGradient::RegisterNatives(g_natives);
-    CanvasImage::RegisterNatives(g_natives);
-    CanvasPath::RegisterNatives(g_natives);
-    CanvasPathMeasure::RegisterNatives(g_natives);
-    Codec::RegisterNatives(g_natives);
-    ColorFilter::RegisterNatives(g_natives);
-    DartRuntimeHooks::RegisterNatives(g_natives);
-    EngineLayer::RegisterNatives(g_natives);
-    FontCollection::RegisterNatives(g_natives);
-    FragmentProgram::RegisterNatives(g_natives);
-    ImageDescriptor::RegisterNatives(g_natives);
-    ImageFilter::RegisterNatives(g_natives);
-    ImageShader::RegisterNatives(g_natives);
-    ImmutableBuffer::RegisterNatives(g_natives);
-    IsolateNameServerNatives::RegisterNatives(g_natives);
-    NativeStringAttribute::RegisterNatives(g_natives);
-    Paragraph::RegisterNatives(g_natives);
-    ParagraphBuilder::RegisterNatives(g_natives);
-    Picture::RegisterNatives(g_natives);
-    PictureRecorder::RegisterNatives(g_natives);
-    Scene::RegisterNatives(g_natives);
-    SceneBuilder::RegisterNatives(g_natives);
-    SemanticsUpdate::RegisterNatives(g_natives);
-    SemanticsUpdateBuilder::RegisterNatives(g_natives);
-    Vertices::RegisterNatives(g_natives);
-    PlatformConfiguration::RegisterNatives(g_natives);
+#define FFI_METHOD_MATCH(CLASS, METHOD, ARGS)                   \
+  if (strcmp(name, #CLASS "::" #METHOD) == 0 && args == ARGS) { \
+    return reinterpret_cast<void*>(                             \
+        tonic::FfiDispatcher<CLASS, decltype(&CLASS::METHOD),   \
+                             &CLASS::METHOD>::Call);            \
   }
+
+void* ResolveFfiNativeFunction(const char* name, uintptr_t args) {
```

Comment:
Can this be a hash table lookup instead?

Reply:
Also, does this get cached by the Dart VM, or do we pay this lookup cost on every call?

Reply:
I did a test implementation of map-based lookup a while ago while reviewing the original version of this PR.  I plan to revive it after this PR lands.

The lookup is cached by Dart and is only done once for each native function.